# Exercise 01 — Basic Operations

In [79]:
import pandas as pd

In [80]:
import pandas as pd

views = pd.read_csv(
    'data/feed-views.log',
    sep='\t',
    skiprows=[2, 3],
    skipfooter=2,
    engine='python',
    names=['datetime', 'user']
)

#  datetime ni to'g'ri dtype ga o'tkazish
views['datetime'] = pd.to_datetime(views['datetime'])



In [81]:
#  yangi columnlar chiqarish
views['year'] = views['datetime'].dt.year
views['month'] = views['datetime'].dt.month
views['day'] = views['datetime'].dt.day
views['hour'] = views['datetime'].dt.hour
views['minute'] = views['datetime'].dt.minute
views['second'] = views['datetime'].dt.second

views.head()

,datetime,user,year,month,day,hour,minute,second
0,2020-04-17 12:01:08.463179,artem,2020,4,17,12,1,8
1,2020-04-17 12:01:23.743946,artem,2020,4,17,12,1,23
2,2020-04-17 12:35:52.735016,artem,2020,4,17,12,35,52
3,2020-04-17 12:36:21.401412,oksana,2020,4,17,12,36,21
4,2020-04-17 12:36:22.023355,oksana,2020,4,17,12,36,22


In [82]:
#daytime ni aniqlash

bins = [-1, 3, 6, 10, 16, 19, 23]
labels = [
    'night',
    'early morning',
    'morning',
    'afternoon',
    'early evening',
    'evening'
]

views['daytime'] = pd.cut(views['hour'], bins=bins, labels=labels)

In [83]:
#User ustunini indexga aylantirish
views.set_index('user', inplace=True)

In [84]:
#elementlari soni
total_count = views.count()
total_count

datetime    1072
year        1072
month       1072
day         1072
hour        1072
minute      1072
second      1072
daytime     1072
dtype: int64

In [85]:
#qiymati bo'yicha vaqtni sortlash
views_sorted = views.sort_values(by=['hour', 'minute', 'second'])


In [86]:
# min / max / mode 
min_hour = views['hour'].min()
max_hour = views['hour'].max()
mode_daytime = views['daytime'].mode()

In [87]:
night_max = views[views['daytime'] == 'night']['hour'].max()

In [88]:
morning_df = views[views['daytime'] == 'morning']

morning_min = morning_df['hour'].min()
morning_user = morning_df[morning_df['hour'] == morning_min].index[0]

In [89]:
mode_hour = views['hour'].mode()
mode_daytime = views['daytime'].mode()

In [90]:
earliest = views.nsmallest(3, 'hour')[['hour']]
latest = views.nlargest(3, 'hour')[['hour']]

earliest
latest

,hour
user,
konstantin,23
artem,23
artem,23


In [91]:
desc = views.describe()
desc

,datetime,year,month,day,hour,minute,second
count,1072,1072.0,1072.000000,1072.000000,1072.000000,1072.000000,1072.000000
mean,2020-05-10 09:30:16.241518,2020.0,4.872201,13.531716,16.256530,29.602612,29.508396
min,2020-04-17 12:01:08.463179,2020.0,4.000000,1.000000,0.000000,0.000000,0.000000
25%,2020-05-10 01:13:53.448644,2020.0,5.000000,11.000000,13.000000,14.000000,14.000000
50%,2020-05-11 22:48:35.302553,2020.0,5.000000,13.000000,19.000000,29.000000,30.000000
75%,2020-05-14 14:30:41.062071,2020.0,5.000000,14.000000,22.000000,46.000000,45.000000
max,2020-05-21 23:23:49.995349,2020.0,5.000000,30.000000,23.000000,59.000000,59.000000
std,NaN,0.0,0.334021,4.901354,6.960369,17.710471,17.424398


In [92]:
iqr = desc.loc['75%', 'hour'] - desc.loc['25%', 'hour']
iqr

np.float64(9.0)

# 📊 Exercise 01 — Data Analysis Report

## 1. Dataset Overview
The dataset contains **1072 records** of user activity logs.

- No missing values detected
- Data spans approximately **1 month**
- Columns include datetime and extracted time features

---

## 2. Time Distribution Analysis

### Key Statistics:
- Mean hour: **16**
- Median hour: **19**
- Min hour: **0**
- Max hour: **23**

### Interpretation:
Most users visit the platform during the **afternoon and evening**.

---

## 3. Quartile Analysis

- Q1 (25%): **13:00**
- Median (50%): **19:00**
- Q3 (75%): **22:00**

### Interpretation:
- 25% of visits happen before 13:00  
- 50% before 19:00  
- 75% before 22:00  

👉 Majority of activity is concentrated in the **second half of the day**

---

## 4. Interquartile Range (IQR)

- IQR = Q3 - Q1 = **22 - 13 = 9**

### Interpretation:
The central 50% of user activity occurs between:

**13:00 and 22:00**

👉 This indicates strong activity during **afternoon → evening hours**

---

## 5. Daytime Behavior

Based on categorized time intervals:

- Afternoon and evening dominate traffic
- Night and early morning activity is minimal

---

## 6. User Behavior Insights

- Users are most active after midday
- Peak usage occurs between **19:00–22:00**
- Activity is consistent across late hours

---

## 7. Data Quality

- No missing values
- Reasonable distribution of time features
- Limited time span (~1 month)

---

## 8. Conclusion

User activity is clearly **time-dependent**:

- Low activity in early hours
- Increasing activity after noon
- Peak in evening

👉 This suggests users engage with the platform mostly **after daily routines (work/study)**

---

## 9. Recommendation (Optional Insight)

If this were a real product:

- Best time for updates → morning
- Best time for engagement → evening
- Marketing campaigns → after 17:00